### PR TITLE
Ruby 2.6: String#split with block

### LINKED ISSUE #2050: String#split with block incompatibility
MRI behavior:

```ruby
"a b".split(" ", &proc{})
#=> "a b"
```

TruffleRuby (head):

```ruby
"a b".split(" ", &proc{})
#=> ["a", "b"]
```

Looks like this change from 2.6 hasn't been implemented: https://rubyreferences.github.io/rubychanges/2.6.html#stringsplit-with-block (though TruffleRuby "says" its RUBY_VERSION is 2.6).

This behaviour is covered by [ruby/spec](https://github.com/oracle/truffleruby/blob/master/spec/ruby/core/string/split_spec.rb) but it seems that these tests are disabled.

### PR DESCRIPTION
**Pull request by ssnickolay**
Closes #2050 

I had to change almost all `Splitter` to not create array when block is passed. Inspired by MRI implementation: https://github.com/ruby/ruby/blob/1020e120e060c00eca456d5a129c344daa472407/string.c#L7933-L7963

### CODE REVIEW

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 1]
```diff
@@ -3179,19 +3179,21 @@ protected DynamicObject stringAppend(DynamicObject string, DynamicObject other)
...
+        @Child private YieldNode yieldNode = YieldNode.create();
         @Child RopeNodes.CodeRangeNode codeRangeNode = RopeNodes.CodeRangeNode.create();
         @Child private RopeNodes.GetCodePointNode getCodePointNode = RopeNodes.GetCodePointNode.create();
         @Child private SubstringNode substringNode = SubstringNode.create();
 
         private static final int SUBSTRING_CREATED = -1;
 
         @Specialization(guards = "is7Bit(string, codeRangeNode)")
-        protected DynamicObject stringAwkSplitSingleByte(DynamicObject string, int limit,
+        protected DynamicObject stringAwkSplitSingleByte(DynamicObject string, int limit, Object block,
```
Comment (bjfish) [inline src/main/java/org/truffleruby/core/string/StringNodes.java:3190]: I think the implementations are different enough that this could be split into multiple specializations:
`protected DynamicObject stringAwkSplitSingleByteNoBlock(DynamicObject string, int limit, Nil block, ...`
`protected DynamicObject stringAwkSplitSingleByte(DynamicObject string, int limit, DynamicObject block, ...`
Thoughts @eregon ?

File: src/main/ruby/truffleruby/core/splitter.rb [hunk 2]
```diff
@@ -36,157 +36,225 @@
...
         end
       end
 
-      ret
+      def empty_string(original)
+        # Use #byteslice because it returns the right class and taints automatically.
+        original.byteslice(0,0)
+      end
+
+      def result(string, res, &block)
```
Comment (bjfish) [inline src/main/ruby/truffleruby/core/splitter.rb:253]: Style wise, I think we usually lean towards inlining small helper methods.

File: spec/ruby/core/string/split_spec.rb [hunk 3]
```diff
@@ -426,12 +426,62 @@
...
+        a = []
+        returned_object = "chunky".split(//) { |str| a << str.capitalize }
 
-      returned_object.should == "chunky bacon"
-      a.should == ["Chunky", "Bacon"]
+        returned_object.should == "chunky"
+        a.should == %w(C H U N K Y)
+      end
+
+      it "yields each split substrings with empty regexp pattern and limit" do
```
Comment (bjfish) [inline spec/ruby/core/string/split_spec.rb:462]: Typo, this should be singular I think: `substring`.

File: src/main/ruby/truffleruby/core/splitter.rb [hunk 4]
```diff
@@ -36,157 +36,225 @@
...
+          else
+            lim = 0
+          end
+
+          return Primitive.string_awk_split string, lim, block
+        elsif pattern.kind_of?(Regexp)
+          # Handle SPLIT_TYPE_REGEXP below
         else
-          lim = 0
+          pattern = StringValue(pattern) unless pattern.kind_of?(String)
```
Comment (bjfish) [inline src/main/ruby/truffleruby/core/splitter.rb:80]: It should be okay to omit `unless  pattern.kind_of?(String)` as `StringValue` has a `String` check.

File: src/main/ruby/truffleruby/core/splitter.rb [hunk 5]
```diff
@@ -36,157 +36,225 @@
...
+          limit = Primitive.rb_to_int limit
+
+          if limit > 0
+            return [string.dup] if limit == 1
+            limited = true
+          else
+            if limit < 0
+              tail_empty = true
+            end
+            limited = false
```
Comment (bjfish) [inline src/main/ruby/truffleruby/core/splitter.rb:60]: Would it be possible to combine `lim`, `limited`, `limit` to just be `limit` where `limit == nil` represents when it is not provided. It's okay to re-assign the  `limit` variable.

File: src/main/ruby/truffleruby/core/splitter.rb (hunk 2 above)
Comment (ssnickolay) [reply src/main/ruby/truffleruby/core/splitter.rb:253]: It just for DRY but I can inline it if necessary

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 1 above)
Comment (ssnickolay) [reply src/main/java/org/truffleruby/core/string/StringNodes.java:3190]: I thought about it, but doesn't it require 4 methods in the end (because of guards)?

```java
@Specialization(guards = "is7Bit(string, codeRangeNode)")
protected methodWithBlock ...

@Specialization(guards = "!is7Bit(string, codeRangeNode)")
protected methodWithBlock ... 

@Specialization(guards = "is7Bit(string, codeRangeNode)")
protected methodNoBlock ... 

@Specialization(guards = "!is7Bit(string, codeRangeNode)")
protected methodNoBlock ...
```

File: src/main/ruby/truffleruby/core/splitter.rb (hunk 5 above)
Comment (ssnickolay) [reply src/main/ruby/truffleruby/core/splitter.rb:60]: Hm, let's try (I gonna try tomorrow) ... I didn't touch this Rubinius logic on purpose, but maybe it's time to reorganize / simplify this

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/string/StringNodes.java:3190]: There is quite a bit of logic in the two specializations, so duplicating the bodies seem unfortunate.
How about profiling whether a block is passed with a ConditionProfile instead?
That way only one the two branches of addSubstring() will be compiled when always used the same way (always with block, or always without block, taking splitting into account).

File: src/main/ruby/truffleruby/core/splitter.rb (hunk 5 above)
Comment (ssnickolay) [reply src/main/ruby/truffleruby/core/splitter.rb]: There was good advice (also found missed edge case and added test for it): https://github.com/oracle/truffleruby/pull/2052/commits/04fe83538b7604362f76f186aca95cee8db16411

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 1 above)
Comment (ssnickolay) [reply src/main/java/org/truffleruby/core/string/StringNodes.java:3190]: Done: https://github.com/oracle/truffleruby/pull/2052/commits/7c73068d1f4f9f37f3c767e3ef0f3766fd3ff26e

**Comment by ssnickolay** [comment]
@bjfish I see you ran Oracle CI...FYI I've not added a note to CHANGELOG (my bad, sorry)

**Comment by bjfish** [comment]
@ssnickolay There appears to be a failure when running `jt test test/mri/tests/csv/parse/test_quote_char_nil.rb`. 

I started to debug this and have some work in progress here: https://gist.github.com/bjfish/36720052f4a9fbf08d95452cda0b69ac

If you would like to make progress on this, that would be great. Otherwise I can take a look again tomorrow (9-5 US Central time).

**Comment by ssnickolay** [comment]
Don't worry, I'm on it

File: src/main/ruby/truffleruby/core/splitter.rb [hunk 6]
```diff
@@ -36,157 +36,219 @@
...
-      ret << string.byteslice(pos, str_size - pos)
+      def add_empty(string, array, count, &block)
+        count.times { add_or_call(array, empty_string(string), &block) }
+      end
 
-      if trim_end
-        while s = ret.at(-1) and s.empty?
-          ret.pop
+      def add_or_call(array, element, &block)
+        if block_given?
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/splitter.rb:231]: Minor but I'd use `if block` here. In TruffleRuby `if block` vs `if block_given?` (and not having a variable for the block) has no performance difference once compiled. In interpreter, `if block_given?` needs to do a method lookup once to make sure it's `Kernel#block_given?` (and relies on AST-inlining otherwise it's expensive as it needs access to the caller frame).

File: src/main/ruby/truffleruby/core/splitter.rb (hunk 2 above)
Comment (eregon) [reply src/main/ruby/truffleruby/core/splitter.rb:247]: I'd tend to inline it too here here, and express it as a simple ternary condition `block ? string : res`.

**Comment by ssnickolay** [comment]
@bjfish I found the problem: in the PR I replaced `match_from`
https://github.com/oracle/truffleruby/blob/137f908b93ad76a561dffdd6a133edd5c132c044/src/main/ruby/truffleruby/core/regexp.rb#L233-L236
with existing `RegexpOperation#match`

https://github.com/oracle/truffleruby/blob/f6df5389ea0e72df08cba78b4a2ef9d4572235ae/src/main/ruby/truffleruby/core/truffle/regexp_operations.rb#L13-L22

which calls the same `search_region` in the end, but has `pos = Primitive.string_byte_index_from_char_index(str, pos)`.  `string_byte_index_from_char_index` changes `pos` var when string encoding is `UTF-16` (`0 -> 0, 4 -> 8, ...`) and then we get an exception.

I could skip this `pos` modification (for example 
```ruby
def match(re, str, pos=0, use_byte_idx=true)
  ...
  pos = Primitive.string_byte_index_from_char_index(str, pos) if use_byte_idx
  ...
end
```
, but I'm not sure that it is right solution. wdyt?

File: src/main/ruby/truffleruby/core/splitter.rb (hunk 6 above)
Comment (ssnickolay) [reply src/main/ruby/truffleruby/core/splitter.rb]: Good point, done :+1:

Regarding `result` inlining - @bjfish has already suggested this before, so 2vs1, is done too :)

**Comment by eregon** [comment]
Could you keep using `match_from`? Or do you need some logic from `match`?
I think a single method shouldn't mix byte and character indicies, and we should add a comment on those methods saying which indices it uses.

**Comment by ssnickolay** [comment]
> Could you keep using match_from? Or do you need some logic from match?
I think a single method shouldn't mix byte and character indicies, and we should add a comment on those methods saying which indices it uses.

My goal was (and is) to remove `match_from` from the Regexp class because there is no the same method in MRI. So, I yes, can add `match_from` (keep this name?) to `RegexpOperation` and use it from there.

**Comment by eregon** [comment]
On a general note, I wonder if it's worth skipping the array here, as it adds a lot of extra logic.
It's just one allocation + growing as needed, and it might escape analyse (if `split` is split/cloned for usages using a block).
The implementation would be trivial if we just call the block for each array element.
Now that the work is done let's use it, but it's something to keep in mind: allocations might be expensive on CRuby, but not necessarily on TruffleRuby (one would need to benchmark to know).

Maybe another way here would be to internally always use the block version, and when the user passes no block, pass a block appending to an array and return that array. Then we don't have so many conditions inside the implementation for "block or not". It's a little bit of overhead for the non-block version, but I'd think it'd be acceptable.

**Comment by eregon** [comment]
Yes, let's mpve `match_from` to RegexpOperations.

**Comment by ssnickolay** [comment]
> The implementation would be trivial if we just call the block for each array element.
Now that the work is done let's use it, but it's something to keep in mind: allocations might be expensive on CRuby, but not necessarily on TruffleRuby (one would need to benchmark to know).

I abandoned this idea in the first place, since the meaning of using a split+block is a predictable large number of splitting results (a huge final array). You are right, I didn't bench how significant the effect on memory/time will be, if we always collect the array, and then either execute the block or return it as it is...I just thought if there is an opportunity to do it optimized, then why not use it :)

**Comment by eregon** [comment]
Yeah, that's the motivation for CRuby to add it. Indeed intuitively a very large array would have some overhead (unclear to me if significant in time).
I see it as a trade-off of maintainability and performance. And branches in the inner loop are not ideal.

Could you benchmark on a large String with many splits how does `str.split(sep)` vs `ary=[]; str.split(sep) { |e| ary << e }` perform? If they are close enough, I think the alternative to always use a block internally would be nicer for maintainability/readability.

**Comment by ssnickolay** [comment]
### The code
```ruby
require 'benchmark'

string = "a:" * 500_000
n = # ???
Benchmark.bm do |x|
  x.report("default          ") { n.times do string.split(":") end}
  x.report("with block       ") { n.times do arr = []; string.split(":") { |e| arr << e } end }
  x.report("with empty block ") { 1.upto(n) do  string.split(":", &Proc.new {}) end }
end
```

### The Results (includes an MRI to see the difference):

**n = 1**
```ruby
# MRI

       user     system      total        real
default           0.048539   0.004992   0.053531 (  0.053554)
with block        0.079062   0.004594   0.083656 (  0.083754)
with empty block  0.053232   0.000375   0.053607 (  0.053710)

# TruffleRuby

       user     system      total        real
default           5.066197   0.216698   5.282895 (  0.620385)
with block        6.329772   0.248593   6.578365 (  1.026183)
with empty block  1.700088   0.045460   1.745548 (  0.345881)
```

**n = 5_000**
```ruby
# MRI

       user     system      total        real
default           199.724127   2.085504 201.809631 (201.917357)
with block        289.136842   2.134888 291.271730 (291.414768)
with empty block  213.914942   0.113967 214.028909 (214.146433)

# TruffleRuby

       user     system      total        real
default          59.624243   2.506349  62.130592 ( 48.558280)
with block       66.996854   0.585985  67.582839 ( 59.105438)
with empty block 36.002969   0.083204  36.086173 ( 33.678120)
```

Seems we could use `block ||= proc { |s| arr << s }` ... what about `StringNodes#string_awk_split`? `Object[] ret = new Object[10]` creates "on java side" for this case

**Comment by eregon** [comment]
Could you run it with TruffleRuby's harness like:
```
jt -u jvm-ce benchmark -e 'string = "a:" * 500_000; benchmark { string.split(":") }'
```
This is simple and then ensures each benchmark is run in isolation.

Alternatively you can run it with `benchmark/ips` so warmup is accounted for and there is no need to hardcode `n`.
`x.iterations = 2` should be used so we see if it's stable, and we make sure all variants see the same polymorhism.
It's not convenient to run them in isolation that way though, so here the first approach seems best.

**Comment by ssnickolay** [comment]
Here are numbers: https://gist.github.com/ssnickolay/249ca69305c0a51b5ff1ef48b0340687

**Comment by eregon** [comment]
OK, so `string.split(":")` is significantly faster than `arr = []; string.split(":") {|e| arr << e}`.
I tried a variant:
```
jt -u jvm-ce benchmark -e 'string = "a:" * 500_000; benchmark { string.split(":") }'
...
84.74798005614377

jt -u jvm-ce benchmark -e 'string = "a:" * 500_000; benchmark { arr = []; string.split(":") {|e| arr << e} }'
...
52.10164807502156

jt -u jvm-ce benchmark -e 'string = "a:" * 500_000; benchmark { arr = []; string.split(":", &arr.method(:<<)) }'       
...
87.3758197450841
```
So the literal block version seems slow, probably because of the extra indirection and needing to read the `arr` captured variable which is many calls above and probably not part of the compilation.
But `&arr.method(:<<)` is fast, because it reads the receiver just from the `Method` object, and doesn't have the extra block as indirection.

Also we see `string.split(":") { |e| }` is slower than `string.split(":")`, so now I'm curious if `string.split(":").each { |e| }` is faster than `string.split(":") { |e| }`.
And it is for me locally (76 for the `each` version vs 64).
But OTOH that uses a higher memory footprint (4-8x the string length bytes at worse, still a lot).

So I think we could always use a block, and when a block is not given by the user, use `&arr.method(:<<)` as the block.
That should improve readability, doesn't seem to lose any performance, and keeps memory footprint efficient.

**Comment by eregon** [comment]
For the `string_awk_split` the results are different though, the version without a block seems much faster:
```
jt -u jvm-ce benchmark -e 'string = "a " * 500_000; benchmark { string.split() }'
19.729362859557416
52.23885591101395
56.78504099610921
58.89106100387889
59.51064425951547
58.90381737029247
61.39174614310269
60.264946697438084
60.45569800304582
61.034904421419455
60.8671225672849
61.50391368624721

jt -u jvm-ce benchmark -e 'string = "a " * 500_000; benchmark { arr = []; string.split {|e| arr << e} }' 
5.721410996060692
9.33220518420548
15.930083246954764
18.61557001135991
18.8023410364349
18.941676527012376
18.94391803564731
18.973680722019353
18.88626699355253
18.889732130036514
19.241663445349708
19.07720313274179
19.109229733014896

jt -u jvm-ce benchmark -e 'string = "a " * 500_000; benchmark { arr = []; string.split(&arr.method(:<<)) }'
6.254126003305206
15.480062552773223
24.549281573107383
26.716961495094772
26.415700712226247
27.18849065513795
26.807189690263176
27.360731794176033
27.23894091583156
27.06189556161548
27.26768768070673
27.626364438765297
27.37573588980818

jt -u jvm-ce benchmark -e 'string = "a " * 500_000; benchmark { arr = []; string.split {|e|} }'   
8.282393918914574
25.839473626665484
27.99681085863714
28.094446382348515
28.677861009605646
29.493731006353393
29.571179847363265
29.278615480585312
29.464236895997654
29.566116958358766
29.47101420724248
29.62411066158248
```

So for `string_awk_split` let's leave both variants in I guess (as in current PR).

**Comment by ssnickolay** [comment]
@eregon thanks for the good idea about the default block. I like that we were able to achieve two goals at once - good performance and maintainability. 

FYI I cheched (without block vs `&arr.method(:<<)` ) other `String#split` edge cases and for all of them I got the same result as yours.

Probably the most interesting was this one:

```
jt -u jvm-ce benchmark -e 'string="a" * 500_000; benchmark { string.split("", 100) }'

3.4789628096894316
7.2330344487933935
11.523495171039034
13.342410198685933
18.12851613615416
18.263387339760136
18.68208280638849
18.594560956532163
18.689901026792082
18.354406277737645
19.155632388140738
19.131784822035204
19.038159630829384

jt -u jvm-ce benchmark -e 'string="a" * 500_000; benchmark { arr = []; string.split("", 100, &arr.method(:<<)) }'

74.40797280483093
509.3267430601462
1800.3935697155155
6465.042696287435
10296.733757984048
34814.85262773839
70021.2467423125
192266.04591950463
297243.4379123956
278637.9738387349
279909.2665481963
303596.3936352284
301716.1844629922
307914.2831240885
302739.64869529416
308307.30870383576
```

AFAIU we've improved this case up to x15k times (because we dont take limit from _whole_ result array) :o

**Comment by eregon** [comment]
Thank you for another great PR!
